WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -83,18 +85,42 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
   NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  if (nil == matchingSnapshots) {
+    return @[];
+  }
+  // Prefiltering elements speeds up search by XPath a lot, because [element resolve] is the most expensive operation here
+  NSDictionary *elementsByType = [self filterPotentialMatchesByType:matchingSnapshots];
+  NSArray *matchingElements = [self filterElements:elementsByType matchingSnapshots:matchingSnapshots];
   return matchingElements;
 }
 
-- (NSArray<XCUIElement *> *)filterElements:(NSArray<XCUIElement *> *)elements matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
+- (NSDictionary<NSNumber *, NSArray<XCElementSnapshot *> *> *)filterPotentialMatchesByType:(NSArray<XCElementSnapshot *> *)snapshots
```

Comment:
This will return mapping from type to array of `XCUIElement` not `XCElementSnapshot`. 
Also method name is a bit confusing. You should consider splitting this method into two steps/methods:
1) get snapshot types from `matchingSnapshots`. Key value coding for array gives you that.
2) create typeToDescendantElementMapping:forTypes:

It will be more readable and will give better testability.